pkg/providers/subnet/subnet.go

```diff
@@ -103,58 +116,64 @@ func (p *DefaultProvider) List(ctx context.Context, nodeClass *v1beta1.EC2NodeCl
 
 // CheckAnyPublicIPAssociations returns a bool indicating whether all referenced subnets assign public IPv4 addresses to EC2 instances created therein
 func (p *DefaultProvider) CheckAnyPublicIPAssociations(ctx context.Context, nodeClass *v1beta1.EC2NodeClass) (bool, error) {
-	subnets, err := p.List(ctx, nodeClass)
-	if err != nil {
-		return false, err
+	for _, subnet := range nodeClass.Status.Subnets {
+		if subnetAssociatePublicIP, ok := p.associatePublicIPAddressCache.Get(subnet.ID); ok && subnetAssociatePublicIP.(bool) {
+			return true, nil
+		}
 	}
-	_, ok := lo.Find(subnets, func(s *ec2.Subnet) bool {
-		return aws.BoolValue(s.MapPublicIpOnLaunch)
-	})
-	return ok, nil
+	return false, nil
 }
 
 // ZonalSubnetsForLaunch returns a mapping of zone to the subnet with the most available IP addresses and deducts the passed ips from the available count
-func (p *DefaultProvider) ZonalSubnetsForLaunch(ctx context.Context, nodeClass *v1beta1.EC2NodeClass, instanceTypes []*cloudprovider.InstanceType, capacityType string) (map[string]*ec2.Subnet, error) {
-	subnets, err := p.List(ctx, nodeClass)
-	if err != nil {
-		return nil, err
-	}
-	if len(subnets) == 0 {
+func (p *DefaultProvider) ZonalSubnetsForLaunch(ctx context.Context, nodeClass *v1beta1.EC2NodeClass, instanceTypes []*cloudprovider.InstanceType, capacityType string) (map[string]*Subnet, error) {
+	if len(nodeClass.Status.Subnets) == 0 {
 		return nil, fmt.Errorf("no subnets matched selector %v", nodeClass.Spec.SubnetSelectorTerms)
 	}
+
 	p.Lock()
 	defer p.Unlock()
-	// sort subnets in ascending order of available IP addresses and populate map with most available subnet per AZ
-	zonalSubnets := map[string]*ec2.Subnet{}
-	sort.Slice(subnets, func(i, j int) bool {
-		iIPs := aws.Int64Value(subnets[i].AvailableIpAddressCount)
-		jIPs := aws.Int64Value(subnets[j].AvailableIpAddressCount)
-		// override ip count from ec2.Subnet if we've tracked launches
-		if ips, ok := p.inflightIPs[*subnets[i].SubnetId]; ok {
-			iIPs = ips
+
+	zonalSubnets := map[string]*Subnet{}
+	availableIPAddressCount := map[string]int64{}
+	for _, subnet := range nodeClass.Status.Subnets {
+		if subnetAvailableIP, ok := p.availableIPAddressCache.Get(subnet.ID); ok {
+			availableIPAddressCount[subnet.ID] = subnetAvailableIP.(int64)
 		}
-		if ips, ok := p.inflightIPs[*subnets[j].SubnetId]; ok {
-			jIPs = ips
+	}
+
+	for _, subnet := range nodeClass.Status.Subnets {
+		if v, ok := zonalSubnets[subnet.Zone]; ok {
+			currentZonalSubnetIPAddressCount := v.AvailableIPAddressCount
+			newZonalSubnetIPAddressCount := availableIPAddressCount[subnet.ID]
+			if ips, ok := p.inflightIPs[v.ID]; ok {
+				currentZonalSubnetIPAddressCount = ips
+			}
+			if ips, ok := p.inflightIPs[subnet.ID]; ok {
+				newZonalSubnetIPAddressCount = ips
+			}
+
+			if currentZonalSubnetIPAddressCount < newZonalSubnetIPAddressCount {
```

Comment:
This statement here on the one on L159 are the same. Is there a way to collapse this into a single line that we call once in either scenario.